lib/LANraragi/Plugin/Metadata/HentagOnline.pm

```diff
@@ -0,0 +1,238 @@
+
+        #Return tags IFF data is found
+        $logger->info("Sending the following tags to LRR: $tags");
+        if ( $save_title && $title ) {
+            $logger->info("Parsed title is $title");
+            return ( tags => $tags, title => $title );
+        } elsif ($tags ne "") {
+            return ( tags => $tags );
+        }
+    }
+
+    return ( error => "No matching Hentag Archive Found!" );
+}
+
+# Returns the ID from a hentag URL, or undef if invalid.
+sub parse_vault_url($url) {
+    if (!defined $url) {
+        return;
+    }
+    if ( $url =~ /https?:\/\/(?:www\.)?hentag\.com\/vault\/([a-zA-Z0-9]*)\/?.*/ ) {
+        return $1;
+    }
+    return;
+}
+
+# Fairly good for mocking in tests
+# get_json_by_title(ua, archive_title, logger)
+sub get_json_by_title($ua, $archive_title, $logger) {
+    my $stringjson = '';
+    my $url = Mojo::URL->new('https://hentag.com/api/v1/search/vault/title');
+    $logger->info("Hentag search for $archive_title");
+
+    my $res = $ua->post($url => json => {title => $archive_title})->result;
+
+    if ($res->is_success) {
+        $stringjson = $res->text;
+        $logger->info('Successful request, response: '.$stringjson);
+    }
+    return $stringjson;
+}
+
+# Fairly good for mocking in tests
+# Returns the json response (as a string) from hentag. If no response, an empty string is returned.
+sub get_json_by_vault_id($ua, $vault_id, $logger) {
+    my $string_json = '';
+    my $url = Mojo::URL->new('https://hentag.com/api/v1/search/vault/id');
+    $logger->info("Hentag search for $vault_id");
+
+    my $res = $ua->post($url => json => {ids => [$vault_id]})->result;
+
+    if ($res->is_success) {
+        $string_json = $res->text;
+        $logger->info('Successful request, response: '.$string_json);
+    }
+    return $string_json;
+}
+
+# Fairly good for mocking in tests
+# Returns the json response (as a string) from hentag. If no response, an empty string is returned.
+sub get_json_by_urls($ua, $logger, @urls) {
+    my $string_json = '';
+    my $url = Mojo::URL->new('https://hentag.com/api/v1/search/vault/url');
+    $logger->info("Hentag search based on URLs: @urls");
+    my $res = $ua->post($url => json => {urls => \@urls})->result;
+
+    if ($res->is_success) {
+        $string_json = $res->text;
+        $logger->info('Successful request, response: '.$string_json);
+    }
+    return $string_json;
+}
+
+# Fetches tags and title, optionally restricted to a language
+sub tags_in_language_from_hentag_api_json($json, $language = undef) {
+    $language =~ s/^\s+|\s+$//g;
+    $language = lc($language);
+
+    # The JSON can contain multiple hits. Loop through them and pick the "best" one.
+    # Possible improvement: Look for hits with "better" metadata (more tags, more tags in namespaces, etc).
+    foreach my $work (@$json) {
+        # Filter out any hits in the wrong language, if requested. Anyone competent in perl, go ahead and golf this shit
```

Comment:
I think this could be improved upon by mapping `@$json` to a hash with the language as the key, and then looking for `$language` in that hash:  

```
my %json_per_lang = map { LANraragi::Plugin::Metadata::Hentag::language_from_hentag_json($_) => $_ } @$json;

if (exists $json_per_lang{$language}) {
   my ( $tags, $title ) = LANraragi::Plugin::Metadata::Hentag::tags_from_hentag_json($json_per_lang{$language});
   return ($tags, $title);
}
```

See https://perldoc.perl.org/functions/map. I think this code works as-is but haven't tested it :^)

Reply:
Aight! It gets a bit trickier as there can be multiple hits for a single language, but it should be solvable.

Reply:
It's a lot better now I think!